Clicking on figure/table links

KarinaMendez2 opened the pull request:
for: https://app.zenhub.com/workspaces/openstax-unified-5b71aabe3815ff014b102258/issues/openstax/unified/805

janKozie1 commented on src/app/components/ContentExcerpt.tsx:
```diff
@@ -0,0 +1,37 @@
+}
+
+// tslint:disable-next-line:variable-name
+const ContentExcerpt = (props: Props) => {
```
I was thinking, maybe we could make `ContentExcerpt` accept a render prop as children instead?
You could then make the signature like

```js
<ContentExcerpt content={highlight.highlightedContent}>
    {(parsedContent) => <HighlightContent dangerouslySetInnerHTML={parsedContent}/> }
</ContentExcerpt>
```

I think it could make the component focus on parsing content and not doing anything where-the-content-is-coming-from specific.

KarinaMendez2 replied:
what about the other props? So you mean like render two components?

janKozie1 replied:
But does the `ContentExcerpt` care about other props? All it needs is html and data needed to parse that html, so book and page. All other props could be directly passed into the child.

KarinaMendez2 replied:
well data-highlight-id seams important, maybe className can always be the same but idk

janKozie1 replied:
But you can pass the other props directly into the child like

```js
<ContentExcerpt content={highlight.highlightedContent}>
    {(parsedContent) => <HighlightContent dangerouslySetInnerHTML={parsedContent} data-highlight-id=""/> }
</ContentExcerpt>
```

ContentExcerpt wouldn't be aware of what you do with the parsed content.

KarinaMendez2 replied:
ok, um what about when the excerpt isn't a highlight?

janKozie1 replied:
Well 

```js
<ContentExcerpt content={someHTML}>
    {(parsedContent) => <SomethingThatRendersParsedHTML otherProps={} dangerouslySetInnerHTML={parsedContent}/> }
</ContentExcerpt>
```

TomWoodward replied:
@janKozie1 it seems like the purpose of your approach would be to pass an element that handles the display styling of the excerpt, where this component would only do the parsing.

i think your idea is reasonable, but i don't see why we couldn't accomplish the same thing at this point by simply extending `ContentExcerpt` as a styled component like:

const HighlightContent = styled(ContentExcerpt)`
  ...special highlight styles for the excerpt
`

which i feel is more straightforward and follows more common patterns.

if we ever run into a situation where this approach isn't working, we may consider coming back to your proposal here

TomWoodward commented on src/app/content/highlights/components/SummaryPopup/HighlightListElement.tsx:
```diff
@@ -3,11 +3,10 @@ import React from 'react';
 import styled, { css } from 'styled-components/macro';
 import { useAnalyticsEvent } from '../../../../../helpers/analytics';
-import { bodyCopyRegularStyle } from '../../../../components/Typography';
+import ContentExcerpt, { HighlightContent } from '../../../../components/ContentExcerpt';
```
why do you need `HighlightContent` here? is it even used?

TomWoodward commented on src/app/content/studyGuides/components/StudyGuidesListElement.tsx:
```diff
@@ -1,11 +1,11 @@
 import React from 'react';
 import styled, { css } from 'styled-components/macro';
-import { bodyCopyRegularStyle, textRegularStyle } from '../../../components/Typography';
+import ContentExcerpt, { HighlightContent } from '../../../components/ContentExcerpt';
```
same

TomWoodward commented on src/app/content/highlights/components/SummaryPopup/HighlightListElement.tsx:
```diff
@@ -154,10 +139,10 @@ const HighlightListElement = ({ highlight, locationFilterId, pageId }: Highlight
     <HighlightContentWrapper color={highlight.color}>
-      <HighlightContent
+      <ContentExcerpt
         className='summary-highlight-content'
```
i believe this class is what applies the content styling. you should move it into `ContentExcerpt` and maybe change the name

KarinaMendez2 replied:
is `summary-excerpt-styles` fine?

TomWoodward replied:
why not `.content-excerpt`?

TomWoodward commented on src/app/content/components/ContentExcerpt.tsx:
```diff
@@ -0,0 +1,36 @@
+  * {
+    overflow: initial;
+  }
+` ;
```
extra space

TomWoodward commented on src/app/content/components/ContentExcerpt.tsx:
```diff
@@ -0,0 +1,36 @@
+  );
+
+  return <div
+    className='content-excerpt'
```
youre losing this to the classname that styled is passing in

KarinaMendez2 replied:
fixed

TomWoodward commented on src/app/content/components/ContentExcerpt.tsx:
```diff
@@ -0,0 +1,37 @@
+    {...excerptProps}
+  />;
+}).attrs({
+  className: 'content-excerpt',
```
pretty sure this is going to break your ability to extend these styles, it would be more straightforward to do 

```
const ContentExcerpt = styled(({content, className, ...otherProps}: Props) => {
  ...
  return <div
    dangerouslySetInnerHTML={{ __html: fixedContent }}
    className={`content-excerpt ${className}`}
    {...otherProps}
  />;
})

TomWoodward reviewed (changes requested):
we're gonna try a rebase workflow on this PR and see how it goes. once you finish addressing my comments please squash all these commits into one for "extract content excerpt logic into a common component" and begin adding your changes for the next stage of the issue.

when rebasing/squashing it is possible to lose changes if you do it wrong, so if you're nervous you may want to create another branch as backup or literally copy the files somewhere so you don't have to stress out about it.

TomWoodward commented on src/app/content/utils/urlUtils.ts:
```diff
@@ -100,3 +100,8 @@ export const toRelativeUrl = (from: string, to: string) => {
 };
+
+export const fromRelativeUrl = (base: string, to: string) => {
+  const urlBase = new URL(to, `https://openstax.org${base}`);
```
can you add a comment here that the hostname is actually irrelevant
```suggestion
  // this hostname is required by the URL constructor but we ignore it in our response, the value is irrelevant
  const urlBase = new URL(to, `https://openstax.org${base}`);
```

Commit:
adding fromRelativeUrl function , tests and require url globally

TomWoodward commented on script/entry.js:
```diff
@@ -9,6 +9,9 @@ const extensions = ['.ts', '.tsx'];
 const JSDOM = require('jsdom').JSDOM;
 global.DOMParser = new JSDOM().window.DOMParser;
 
+const URL = require ('url');
```
extra space around the parens

TomWoodward commented on src/app/content/components/ContentExcerpt.tsx:
```diff
@@ -1,24 +1,44 @@
+    return null;
+  }
+
+  const sourcePage = findArchiveTreeNodeById(currentBook.tree, sourcePageId) as unknown as Page;
```
you shouldn't need this cast, it is bad

TomWoodward commented on src/app/content/components/ContentExcerpt.tsx:
```diff
@@ -1,24 +1,44 @@
   const fixedContent = React.useMemo(
-    () => addTargetBlankToLinks(props.content),
-    [props.content]
+    () => addTargetBlankAndFixRelativeURL(props.content, excerptSource.url),
```
don't combine these into one function. leave the old one the way it was and create a new one for fixing the relative urls.

https://en.wikipedia.org/wiki/Single-responsibility_principle

Commit:
updating ContentExcerpt to accept excerpt source as a prop to build the base url and
use the new fromRelativeUrl function to transform relative urls in the excerpt into full url paths

TomWoodward commented on src/app/content/components/ContentExcerpt.tsx:
```diff
@@ -1,28 +1,52 @@
 
+  const fixedContent = React.useMemo(() => flow(
+    addTargetBlankToLinks,
+    (newContent) => fixRelativeURLs(newContent, excerptSource.url)),
```
add the call for this function here instead of on line 49, otherwise you're not memoizing the logic

TomWoodward commented on src/app/content/components/ContentExcerpt.tsx:
```diff
@@ -1,28 +1,52 @@
+  const sourcePage = findArchiveTreeNodeById(currentBook.tree, sourcePageId);
+
+  if (!sourcePage) {
+    return null;
```
you should throw instead of nooping here

TomWoodward commented on src/app/content/highlights/components/SummaryPopup/HighlightListElement.tsx:
```diff
@@ -121,7 +121,7 @@ const HighlightListElement = ({ highlight, locationFilterId, pageId }: Highlight
 
   const confirmDelete = () => {
-    dispatch(requestDeleteHighlight(highlight, {
+    dispatch(deleteHighlight(highlight, {
```
why are you changing this?

TomWoodward replied:
why are you changing this?

TomWoodward commented on src/app/content/studyGuides/components/__snapshots__/StudyGuides.spec.tsx.snap:
```diff
@@ -43,17 +43,7 @@ exports[`StudyGuides properly display summary highlights 1`] = `
-            data-highlight-id="hl1"
-          />
-        </div>
+        />
```
this seems bad

TomWoodward commented on src/app/content/studyGuides/components/__snapshots__/StudyGuides.spec.tsx.snap:
```diff
@@ -66,17 +56,7 @@ exports[`StudyGuides properly display summary highlights 1`] = `
-            data-highlight-id="hl2"
-          />
-        </div>
+        />
```
and this

TomWoodward commented on src/app/content/utils/fixContentUrls.ts:
```diff
@@ -0,0 +1,28 @@
+import { HTMLAnchorElement } from '@openstax/types/lib.dom';
```
can you rename this file `contentManipulation` or something less url specific

TomWoodward commented on src/app/content/utils/fixContentUrls.spec.ts:
```diff
@@ -47,3 +47,19 @@ describe('addTargetBlankToLinks', () => {
+    const output = '<div class="test">'
+    + '<a href="some/link" target="_blank">First link</a>'
+    + '<a href="some/link" target="_blank">Second link</a>'
+    + '</div>';
```
this test doesn't look like its testing what it says its testing

TomWoodward commented on src/app/content/components/ContentExcerpt.spec.tsx:
```diff
@@ -0,0 +1,67 @@
+
+  beforeEach(() => {
+    store = createTestStore();
+    consoleError = jest
```
this is to ignore the error boundary warning? it would be better to add that to the block list in tests/setup so that any other errors are not hidden

TomWoodward commented on src/app/content/components/ContentExcerpt.spec.tsx:
```diff
@@ -0,0 +1,67 @@
+    </MessageProvider>
+  </Provider>);
+
+  it('fixes urls in content and matches snapshot', () => {
```
you're testing a pretty big integration here, it would be better to mock `addTargetBlankToLinks` and `fixRelativeURLs` and assert that they are called rather than testing what they do

TomWoodward commented on src/app/content/components/ContentExcerpt.tsx:
```diff
@@ -1,26 +1,47 @@
 
+  const fixedContent = React.useMemo(() => flow(
+    addTargetBlankToLinks,
+    (newContent) => fixRelativeURLs(newContent, excerptSource.url))(props.content),
```
can you reformat this so the `)(props.content),` is on the next line please

TomWoodward commented on src/app/content/components/ContentExcerpt.tsx:
```diff
@@ -1,26 +1,47 @@
-    [props.content]
+  const currentBook = useSelector(book);
+
+  if (!currentBook) {
```
is there a legitimate situation where book would not be loaded? if not then we can throw here

TomWoodward commented on src/app/content/utils/contentManipulation.spec.ts:
```diff
@@ -1,5 +1,5 @@
-import addTargetBlankToLinks from './addTargetBlankToLinks';
 
+import { addTargetBlankToLinks } from '../utils/contentManipulation';
```
blank line above this?

TomWoodward commented on src/app/content/utils/contentManipulation.ts:
```diff
@@ -0,0 +1,28 @@
+}
+
+export const isAbsoluteUrl = (url: string) => {
+  const pattern = /^(http)?s?:\/\//i;
```
this matches things it shouldn't you want to move the `s?:` into the parentheses with the `http`

you also don't want to treat urls beginning in a slash, like `/books/asdf/pages/qre` as relative, they aren't fully qualified urls because they don't have the domain or protocol, but they are absolute paths

Commit:
extract content excerpt logic into a common component

TomWoodward commented on src/app/content/components/ContentExcerpt.spec.tsx:
```diff
@@ -0,0 +1,68 @@
+      .toHaveBeenCalledWith(htmlWithTargetBlank, '/book/book1/page/testbook1-testpage1-uuid');
+  });
+
+  it('returns null when book is not loaded', () => {
```
the name of this test is wrong

TomWoodward commented on src/app/content/components/ContentExcerpt.spec.tsx:
```diff
@@ -0,0 +1,68 @@
+    }).toThrowError(new Error('book not loaded'));
+  });
+
+  it('returns null when page is not loaded', () => {
```
and this one

TomWoodward commented on src/test/setup.ts:
```diff
@@ -27,12 +27,18 @@ const ignoreConsoleErrorMessages = [
 
 const originalConsoleError = console.error;  // tslint:disable-line:no-console
-console.error = (msg: unknown) => {  // tslint:disable-line:no-console
+console.error = (msg: string) => {  // tslint:disable-line:no-console
```
your types in here are messed up. its possible to console.error something that isn't a string, and the conditional below does a typecheck on it to see if it is a string. if it makes life easier i'm not opposed to changing the `ignoreConsoleErrorMessages` array to all be regexes

TomWoodward commented on src/app/content/components/ContentExcerpt.tsx:
```diff
@@ -1,26 +1,43 @@
+  const fixedContent = React.useMemo(() => flow(
+    addTargetBlankToLinks,
+    (newContent) => fixRelativeURLs(newContent, excerptSource.url)
+  )(props.content),
+  [props.content, excerptSource.url]);
```
```suggestion
)(props.content), [props.content, excerptSource.url]);
```

TomWoodward reviewed (changes requested):
some minor comments, the logic is all good we're just doing cleanup now

TomWoodward commented on src/app/content/components/ContentExcerpt.tsx:
```diff
@@ -1,26 +1,42 @@
+  const fixedContent = React.useMemo(() => flow(
+    addTargetBlankToLinks,
+    (newContent) => fixRelativeURLs(newContent, excerptSource.url)
+)(props.content), [props.content, excerptSource.url]);
```
indentation is messed up

staxly[bot] committed:
Merge branch 'master' into clicking_links_fix_part1

staxly[bot] committed:
Merge branch 'master' into clicking_links_fix_part1

staxly[bot] committed:
Merge branch 'master' into clicking_links_fix_part1

staxly[bot] committed:
Merge branch 'master' into clicking_links_fix_part1

staxly[bot] committed:
Merge branch 'master' into clicking_links_fix_part1

staxly[bot] committed:
Merge branch 'master' into clicking_links_fix_part1

staxly[bot] committed:
Merge branch 'master' into clicking_links_fix_part1
